### ISSUE TITLE
Feature: Dice rolling

### ISSUE THREAD

**Issue by TwelveBaud** [SockDrawer/SockRPG#22]
As a player or game moderator
I want to be able to reliably and securely "roll dice"
So that I can use chance in my games.

## Dice specification
TBD, but something like
[dice=Initiative]1d20 + 8[/dice]
[dice=Damage Roll]4d8dl + 2d5x + 1dF[/dice]
[dice]d20[/dice]

## Scenario: Basic functionality
Given a post with a dice specification
When the post is posted
Then the dice specification should appear in the post
And additional random numbers, "dice rolls", should appear in the post
And a final value should appear in the post.

Given a post with two dice specifications
When the post is posted
Then both dice specifications, their rolls, and their final values shouldappear in the post
And the numbers and final values should be tied to the dice specification they're from, not others.

Given a post with no dice specifications
And no dice specifications in its history
When the post is posted
Then no dice specification should appear in the post.

Given a post with the dice specification analogous to "[dice]1d20[/dice]"
When the post is posted
Then it should contain something analogous to "1d20: 16"
And that text should be formatted in a distinctive way.

Given a post with the dice specification analogous to "[dice=Damage Roll]1d8 + 2d4 + 6[/dice]"
When the post is posted
Then it should contain something analogous to "Damage Roll: 1d8 [4] + 2d4 [2, 2] + 6 = 14"
And that text should be formatted in a distinctive way.

## Scenario: Finality and Unforgeability
Given a post with a dice specification
And completed rolls for that specification
When the post is edited
And the specification does not change
Then the same rolls should be shown for that specification.

Given a post with a dice specification
And completed rolls for that specification
When the post is edited
And the specification is removed
Then the specification should still be shown to the GM, along with completed rolls.
But the specification should not be shown to other players.

Given a post with a dice specification
And completed rolls for that specification
When the post is edited
And the specification is altered
Then the new specification should get new dice rolls
And the new specification should be shown to other players
And the old specification should be treated as if it was removed.

Given a post with no dice specification
And rolls for a dice specification in the post's history
When the post is edited to add the specification back
Then the specification should appear in the post
And it should have the same rolls it had originally.

Given a post with no dice specification
When the post is posted
Then it should not appear to have a dice specification,
Regardless of the contents of the post.

## Scenario: Attack mitigation
Given a post with a thousand dice specifications
When the post is attempted to be posted
Then the post is rejected

Given a post with ten dice specifications
And they have more than a thousand combined dice
When the post is attempted to be posted
Then the post is rejected

Given a post with ten dice specifications
And they have fewer than 51 combined dice
When the post is attempted to be posted
Then the post is accepted

### Fudge dice
Given a post with a dice specification analogous to "[dice]2dF[/dice]"
When the post is posted
Then it should contain something analogous to "2dF: [+,□] = 1"

### Drop lowest/highest
Given a post with a dice specification analogous to "[dice=Charisma]4d6dl[/dice]"
When the post is posted
Then it should contain something analogous to "Charisma: 4d6dl [3, 2, 1~~, 1~~] = 6"

### Exploding dice
Given a post with a dice specification analogous to "[dice]10d10ww[/dice]"
When the post is posted
And at least one roll is a 10
Then it should contain more than 10 dice rolls.

### Scion dice
Given a post with a dice specification analogous to "[dice]10d10s[/dice]"
When the post is posted
And all rolls are 10
Then the number of successes reported should be 20

### More dice
???

**Comment by Kaelas17** [comment]
**5.0 AD&D Dice**

Given a post with a dice specification analogous to "2a[dice]1d10[/dice]" or 2d[dice]1d10[/dice]"
When the post is posted
Then the it should only present the higher (or lower) of the two results.

(I probably botched how to present this. I'm trying to request the ability for the dice bot to sort out a (dis)advantage roll a la the new Fifth Edition wherein you roll twice and take either the better or worse of the two rolls depending on your circumstances.)

**Comment by AccaliaDeElementia** [comment]
### Rolls should be able to be private
If specified the Roll should be private, only showing the result to the peropn who rolled the dice, Other people can see that a roll was requested, but not what the roll spec was nor what the result was.

(optional?)The GM should be able to "unhide" a hidden roll, making it visible to all players.

**Comment by Kaelas17** [comment]
@AccaliaDeElementia re private rolls, I would suggest that the GM should always be able to see rolls. Dice rolls are meant to represent the flow of fate, luck and chance in the setting, and the setting should always be, if not up to the GM, then at least be made known to the GM. 

I can't think of any instance in which the course of a game should depend on a dice roll that the GM shouldn't be made aware.  Well. From time to time, I've rolled dice to make a choice (left or right path, etc) but those aren't 'real' rolls and shouldn't be in the logs in my opinion.

**Comment by AccaliaDeElementia** [comment]
hmm... yes,

I rather had in mind actually reusing the dice semantics for card draws, in that case there is a big difference between a faceup card and a face down card.

That being said Now that i think of it more closely there are enough differences between dice and cards that they should be separate modules.

**Comment by TwelveBaud** [comment]
> (I probably botched...

Nah, you're doing fine, just need to tuck the "2a" or "2d" inside "[dice]" so that we know it's not part of the rest of the post. The way most other dice programs express that feature is "drop highest" or "drop lowest", by putting "dh" or "dl" on the end, but except when making D&D characters "d" (disadvantage) or "a" (advantage) seems to make more sense to players.

> If specified the roll should be private

Originally I specced it so that the GM could see dice rolls that've been removed from a post, but it could definitely be useful to have them be private from the beginning.

> but those aren't 'real' rolls and shouldn't be in the logs in my opinion.

I thought about that, and I'm concerned people might try to retcon "fake" and "real" rolls into each other. By making all rolls real, and allowing people to tag them ("[dice=Do I go left or right?]") so people know what they're for and whether they're relevant. we can avoid that problem.

**Comment by Kaelas17** [comment]
@AccaliaDeElementia  Huh. Yeah, there are a few systems that might require having a... screw it, Imma just say it. If someone plays Yu-gi-Oh, then they need to be able to legitimately hide their trap cards, even from the GM, until they go off to prevent metagaming (because no-one can perfectly ignore knowledge they shouldn't have but do).

@TwelveBaud Good catch on people trying to retcon rolls. Proper labeling for rolls is very important.

**Comment by AccaliaDeElementia** [comment]
so lets recap what we have, at a very broad level:

We will have multiple dice systems, probably using some sort of plugin/extension system so more can be added by third parties.

Each dice system will have its own dice syntax, and will define a "help" page explaining its syntax that will be linked to in all games that use the dice system.

Dice can be tagged with arbitrary data to indicate what the dice roll is for. GMs may choose to reject rolls that are not tagged with intent but that will not be enforced by the system

Dice rolls can be made private, This will hide the dice spec, tag and the result from all but the player and the GM. All other players will see a message to the effect of `[The sound of rattling dice cna be heard around the table]` to indicate that a roll was made.

Private dice rolls can be revealed by the GM or rolling player. Public rolls cannot be made private.

Rolls will not recalculate on post edit, if the spec changes as part of the edit a new dice roll will be made with the original roll and spec available as history.

Rolls cannot be altered, Only replaced or removed and will always appear in history that they occured.

The GM may trigger a reroll for a non GM character to "fudge" the roll if the GM desires, either letting the roll occur again with a hopefully more favourable result, or by specifying the result of the roll manually. This action will be recorded in history as well and visible on all public rolls.

**Comment by yamikuronue** [comment]
This seems like a HUGE feature. Can we break out each of those scenario sets as a feature maybe?

**Comment by AccaliaDeElementia** [comment]
I was planning to make the base scenario and sample dice rolling (always rolls a 4 regardless of what you ask ti to do) as the base feature and add other features for the individual dice systems.

That seemed a good task breakout to me.
